Add coverage for TestWorker and FetchLinksEmbedData.

derAnfaenger opened the pull request:
This addresses #6542 .

timabbott commented on zerver/worker/queue_processors.py:
```diff
@@ -423,11 +423,11 @@ class TestWorker(QueueProcessingWorker):
+        message = ujson.dumps(event)  # nocoverage
+        logging.info("TestWorker should append this message to %s: %s" % (fn, message))  # nocoverage
+        with open(fn, 'a') as f:  # nocoverage
+            f.write(message + '\n')  # nocoverage
```
You can put `# nocoverage` comments on a block, e.g. the `def consume` block, rather than on every line.  Adding it to every line is messier than just writing a test for the thing :)

timabbott replied:
I merged a version of this change that just touches the `def` line; you can drop this commit.

timabbott commented on zerver/tests/test_link_embed.py:
```diff
@@ -309,3 +309,33 @@ def test_invalid_link(self):
+                    'message_content': "Not the original content 'http://test.org/'."}
+                FetchLinksEmbedData().consume(event)
+        msg = Message.objects.select_related("sender").get(id=msg_id)
+        # If the content of the messaeg has changed since the event has been created, it should not
```
`message` is mispelled.

timabbott commented on zerver/tests/test_link_embed.py:
```diff
@@ -309,3 +309,33 @@ def test_invalid_link(self):
+                    'urls': [url],
+                    'message_realm_id': msg.sender.realm_id,
+                    'message_content': "Not the original content 'http://test.org/'."}
+                FetchLinksEmbedData().consume(event)
```
I'm not sure I understand this test, but I guess I'd have thought that what we want is to actually use the message-editing API...

derAnfaenger replied:
Hm, so I thought what we want to test is this case:
1. Push an event with some message to the FetchLinksEmbedData queue.
2. Before event gets handled, update_message gets called. update_message pushes a new event to the queue.
3. First event gets handled, has outdated message.
4. Event gets dropped
5. Second event gets handled.

This test is incomplete in that it only tests until step 4, and does so by inserting an event with the "outdated" content `Not the original content 'http://test.org/'.`. I agree it would be more useful to have a full flow with 
making a call to `client_patch('message')`, but I didn't see a straightforward way to simulate the racing queue_json_publish calls. I guess I could try smt with mocking and adding side_effects?

timabbott replied:
Yeah, I would just mock `queue_json_publish` for the first thing, capturing the event payload, then do `update_message`.  And then finish as you have it here, call `consume` on the captured event now that the edit has gone through.

derAnfaenger commented:
@timabbott rewrote the test. It should now basically be a complete simulation of the steps outlined in my comment.

derAnfaenger commented:
Added a second commit for having a full test flow through the DigestWorker. The `handle_digest_email` is really heavy and takes around ~0.04 seconds. `test_inactive_users_queued_for_digest` queues 26 inactive users in total, so that amounts to ~0.8 seconds for the test to complete. I think we can limit the test to some percentage of all users?

timabbott commented on zerver/tests/test_link_embed.py:
```diff
@@ -240,6 +241,55 @@ def _send_message_with_test_org_url(self, sender_email, queue_should_run=True, r
+            # The content of the message has changed since the event for original_url has been created,
+            # it should not be rendered. Another, up-to-date event will have been sent (edited_url).
+            self.assertNotIn('<a href="{0}" target="_blank" title="The Rock">The Rock</a>'.format(original_url),
+                             msg.rendered_content)
```
I think maybe a `mocked_response_edited.assert_not_called()` sort of thing would be useful here?

timabbott commented:
Yeah, for Digest emails, I'd only have it runs for the first two, and have the rest return early.

derAnfaenger commented:
Updated. Moved the full-flow tests into `test_active_users_not_enqueued`; the for loop here gradually makes inactive users active, so it's more natural to test the full flow for the last two users. `test_inactive_users_queued_for_digest` on the other hand starts without inactive users only and makes them explicitly inactive.

timabbott commented:
Nice, merged, thanks @derAnfaenger!
